source/slang/slang-ir-specialize-function-call.cpp

```diff
@@ -381,6 +390,58 @@ struct FunctionParameterSpecializationContext
         newCall->insertBefore(oldCall);
         oldCall->replaceUsesWith(newCall);
         oldCall->removeAndDeallocate();
+
+    }
+
+    // After specialization of the function, any resource type parameters will be gone, and they could be replaced as
+    // the index of the resources. For example, if the original function is:
+    //
+    // void foo(Texture2D<float> tex) { ... }
+    //
+    // it could be specialized to:
+    //
+    // void foo(int idx) { ... }
+    //
+    // In such case, we need to check the new arguments of the new call site match the type of the parameters of the
+    // new function, because we didn't do type checking when we generate the new call site. This step is necessary,
+    // because in the old call site, it's allowed to use different data type as the index to retrieve the resources,
+    // however, the new function only allows one data type as the input argument.
+    void typeCastForNewArguments(
```

Comment:
I wonder if we can factor this into the core specialization logic and make the index parameter a part of the specialization key, instead of doing an additional post process step.

Reply:
Yea, I was thinking about that path too, but it takes me more time than implementing this path, and this path seems more straightforward (at least to me). So I just proposed this version of solution.

But since you raised the concern here, I think I can spend more time to consider that refactored path.

Reply:
In the new change, I add the data type of the `index` parameter into the key so we will specialize a new function for this case.